### ISSUE TITLE
Additional energy calibration models

### ISSUE THREAD

**Issue by jccurtis** [lbl-anp/becquerel#99]
Building off of #33, #41, we should add additional energy calibration models using `EnergyCalBase`

**Comment by jccurtis** [comment]
Some partially complete code for a quadratic calibrator I quickly put together:

```python
class QuadraticEnergyCal(energycal.EnergyCalBase):
    """
    kev = a*ch^2 + b*ch + c
    """

    @property
    def valid_coeffs(self):
        """A list of valid coefficients for the calibration curve.
        Returns:
          a tuple of strings, the names of the coefficients for this curve
        """
        return ('a', 'b', 'c')


    def _ch2kev(self, ch):
        """Convert scalar OR np.array of channel(s) to energies.
        Should use numpy ufuncs so that the input dtype doesn't matter.
        Args:
          ch: an np.array, float, or int of channel values
        Returns:
          energy values, the same size/type as ch [keV]
        """
        return self._coeffs['a'] * ch ** 2 + self._coeffs['b'] * ch + self._coeffs['c']

    def _kev2ch(self, kev):
        """Convert energy value(s) to channel(s).
        Args:
          kev: an np.array, float, or int of energy values [keV]
        Returns:
          the channel value(s) corresponding to the input energies.
            a float if input is scalar. an np.array if input is iterable
        """
        # TODO: address with the quadratic inverse formula
        raise NotImplementedError('Sorry')

    def _perform_fit(self):
        """Do the actual curve fitting."""
        a, b, c = np.polyfit(self.channels, self.energies, 2)
        self._set_coeff('a', a)
        self._set_coeff('b', b)
        self._set_coeff('c', c)
```

**Comment by markbandstra** [comment]
We could implement a generic `_kev2ch` in `EnergyCalBase` in the following way, and then allow a specific calibration to override it if there is a faster calculation available:

```python
    def _kev2ch(self, kev):
        kev = np.asarray(kev)
        ch0 = np.arange(10000)  # what is the largest channel we might expect?
        kev0 = self.ch2kev(ch0)
        assert (np.diff(kev0) > 0).all()  # calibration must be monotonically increasing
        assert (kev0.min() <= kev).all() and (kev <= kev0.max())  # make sure we can interpolate
        ch = np.asarray(np.interp(kev, kev0, ch0))
        assert (ch >= 0).all()
        return ch.astype(int)
```

**Comment by jccurtis** [comment]
Nice one, I would alter the last line to return the channel at a `float` to maintain the calibration object's level of precision. Even though fractional channels are undefined in hardware, they are used for accurate calibrations after a peak fit.

**Comment by jccurtis** [comment]
Actually I would go one step further:

```python
    def _kev2ch(self, kev):
        kev = np.asarray(kev, dtype=np.float)
        # Should we allow negative inputs? (This will change the bounds below)
        assert (keV >= 0).all()
        # Find bounds (assumes self._kev2ch(kev) > 0))
        _ch_min = 0.0
        _ch_max = 1.0
       _num_interp_pts = 1e4
        while True:
            _ch_max_kev = self.ch2keV(_ch_max)
            if _ch_max_kev < kev.max():
                if _ch_max_kev < kev.min():
                    _ch_min = _ch_max
                _ch_max *= 2.0
        # Create at vectors of length _num_interp_pts to interpolate from
        if (_ch_max - _ch_min) > _num_interp_pts:
            _ch = np.arange(int(round(_ch_min)), int(round(_ch_max)), dtype=np.float)
        else:
            _ch = np.linspace(_ch_min, _ch_max, _num_interp_pts, dtype=np.float)
        _kev = self.ch2kev(_ch)
        # calibration must be monotonically increasing
        assert (np.diff(_kev) > 0).all()
        # make sure we can interpolate
        assert (_kev.min() <= kev).all() and (kev <= _kev.max())  
        ch = np.asarray(np.interp(kev, kev_interp, _ch))
        # Removing because this won't be allowed by the bound search
        # assert (ch >= 0).all()
        return ch
```

Also regarding `numpy` vs `scipy` for `linear` interpolation. [`numpy` is 5x faster](https://github.com/scipy/scipy/pull/4903#issuecomment-114259888)

**Comment by jccurtis** [comment]
Waiting on #50 ...

**Comment by markbandstra** [comment]
Closed by #248
